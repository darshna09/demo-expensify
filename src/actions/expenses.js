import { v1 as uuidv1 } from 'uuid';

// Action Generator: ADD_EXPENSE
export const addExpense = ({
    description = '', note = '', amount = 0, createdAt = 0
} = {}) => ({
    type: 'ADD_EXPENSE',
    expense: {
        id: uuidv1(),
        description,
        note,
        amount,
        createdAt
    }
});

// Action Generator: REMOVE_EXPENSE
export const removeExpense = (id) => ({
    type: 'REMOVE_EXPENSE',
    id
});

// Action Generator: EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});