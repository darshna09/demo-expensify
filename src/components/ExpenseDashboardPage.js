import React from 'react';

const ExpenseDashboardPage = () => (
    <div>This is from the Expensify Dashboard Page.</div>
);

export default ExpenseDashboardPage;
