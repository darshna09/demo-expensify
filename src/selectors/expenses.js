// Start date, end date  - timestamp (milliseconds)
// Starts from Jnuary 1st 1970 from midnight (UNIX EPOCH)
// Positive milliseconds are after the above time while the ones in negative are before this time.

// Get visible expenses - filtered/sorted
export default (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate;
        const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDate;
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if (sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1;
        } else if (sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1;
        }
    });
};
