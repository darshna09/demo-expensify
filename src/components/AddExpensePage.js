import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.onSubmit(expense);
        this.props.history.push('/');                // It is not a full page refresh but something as the link was pressed
    }

    render() {
        return (
            <div>
                <h1>Add Expense</h1>
                <ExpenseForm onAddExpense={this.onSubmit}/>
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    onSubmit: (expense) => (dispatch(addExpense(expense)))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);

// -------------------------------------------------------------------------------
// For writing tests we need to decouple the dispatch action, as done above.

// const AddExpensePage = (props) => (
//     <div>
//         <h1>Add Expense</h1>
//         <ExpenseForm onAddExpense={(expense) => {
//             props.dispatch(addExpense(expense));
//             props.history.push('/');                // It is not a full page refresh but something as the link was pressed
//         }}/>
//     </div>
// );

// export default connect()(AddExpensePage);
