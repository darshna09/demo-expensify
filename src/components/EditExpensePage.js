import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense } from '../actions/expenses';
import { removeExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
    onEditExpense = (expense) => {
        this.props.onEditExpense(expense);
        this.props.history.push("/");
    }

    onRemoveExpense = () => {
        this.props.onRemoveExpense(this.props.expense.id);
        this.props.history.push("/");
    }

    render() {
        return (
            <div>
                <ExpenseForm
                expense={this.props.expense}
                onAddExpense={this.onEditExpense}
            />
            <button onClick={this.onRemoveExpense}>Remove</button>
            </div>
        );
    }
}

const mapStateToProps = (state, props) => ({
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch) => ({
    onEditExpense: (expense) => dispatch(editExpense(expense.id, expense)),
    onRemoveExpense: (id) => dispatch(removeExpense(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
