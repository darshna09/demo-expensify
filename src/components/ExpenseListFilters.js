import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';

export class ExpenseListFilters extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            focusedInput: null
        }
    }

    onDatesChange = ({startDate, endDate}) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    } 

    setTextFilter = (e) => {
        this.props.setTextFilter(e.target.value);
    }

    sortByFilter = (e) => {
        e.target.value === 'amount' ? this.props.sortByAmount() : this.props.sortByDate();
    }

    render() {
        return (
            <div>
                <input
                    type='text'
                    value={this.props.filters.text}
                    onChange={this.setTextFilter}
                />
                <select
                    value={this.props.filters.sortBy}
                    onChange={this.sortByFilter}
                >
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.focusedInput}
                    onFocusChange={focusedInput => this.setState({ focusedInput })}
                    numberOfMonths={1}
                    showClearDates={true}
                    isOutsideRange={() => false}
                />
            </div>
        )
    }
};

const mapStateToProps = (state) => ({
    filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByAmount: () => dispatch(sortByAmount()),
    sortByDate: () => dispatch(sortByDate()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
