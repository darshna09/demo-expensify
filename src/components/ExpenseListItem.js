import React from 'react';
import { connect } from 'react-redux';
import { removeExpense } from '../actions/expenses';

const ExpenseListItem = ({id, description, amount, createdAt, dispatch}) => (
    <div>
        <p>Description: {description}</p>
        <p>Amount: {amount}</p>
        <p>Created at: {createdAt}</p>
        <button onClick={() => {
            dispatch(removeExpense(id));
        }}>Remove</button>
    </div>
);

export default connect()(ExpenseListItem);