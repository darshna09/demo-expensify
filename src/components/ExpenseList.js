import React from 'react';
import { connect } from 'react-redux';
import getVisibleExpenses from '../selectors/expenses';
import ExpenseListItem from './ExpenseListItem';

const ExpenseList = (props) => (
    <div>
        <h3>Expense List</h3>
        {props.expenses.map((expense) => (<ExpenseListItem key={expense.id} {...expense} />))}
    </div>
);

const mapStateToProps = (state) => ({
    expenses: getVisibleExpenses(state.expenses, state.filters)
});

export default connect(mapStateToProps)(ExpenseList);


// FOR UNDERSTANDING.

// const ConnectedExpenseList = connect((state) => {
//     return {
//         expenses: state.expenses,
//         name: 'Ana'
//     }
// })(ExpenseList);

// export default ConnectedExpenseList;
