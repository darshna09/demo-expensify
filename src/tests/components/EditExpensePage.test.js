import React from 'react';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';
import { EditExpensePage } from '../../components/EditExpensePage';

let onEditExpense, onRemoveExpense, history, wrapper;

beforeEach(() => {
    onEditExpense = jest.fn();
    onRemoveExpense = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(
        <EditExpensePage
            expense={expenses[1]}
            onEditExpense={onEditExpense}
            onRemoveExpense={onRemoveExpense}
            history={history}
        />
    );
});

test('should render EditExpensePage', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should handle edit expense', () => {
    wrapper.find('ExpenseForm').prop('onAddExpense')(expenses[1]);
    expect(onEditExpense).toHaveBeenCalledWith(expenses[1]);
    expect(history.push).toHaveBeenCalledWith('/');
});

test('should handle remove expense', () => {
    wrapper.find('button').simulate('click');
    expect(onRemoveExpense).toHaveBeenCalledWith(expenses[1].id);
    expect(history.push).toHaveBeenCalledWith('/');
});