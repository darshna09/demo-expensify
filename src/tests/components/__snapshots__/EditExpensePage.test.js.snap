// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": "32000",
        "createdAt": -345600000,
        "description": "Credit Card",
        "id": "2",
        "note": "",
      }
    }
    onAddExpense={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
